# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/base/nsGenConList.h
# Commit: 86e9449aba35
# Full Hash: 86e9449aba35c7439e8a41a2804388994462c00d
# Author: Martin Robinson <mrobinson@igalia.com>
# Date: 2022-06-23 03:54:50
# Regressor Bug: 1463600
# File Overlap Count: 1
# Description:
#   Bug 1463600 - Implement CSS 'contain: style' r=emilio
#   
#   Add an implementation of CSS `contain: style`. This introduces two new
#   data structures, the ContainStyleScope and ContainStyleScopeManager.
#   
# ==============================================================================

diff -r c7d1d0cfaa6c -r 86e9449aba35 layout/base/nsGenConList.h
--- a/layout/base/nsGenConList.h	Wed Jun 22 10:30:03 2022 +0000
+++ b/layout/base/nsGenConList.h	Wed Jun 22 10:42:34 2022 +0000
@@ -9,10 +9,12 @@
 #ifndef nsGenConList_h___
 #define nsGenConList_h___
 
+#include "mozilla/FunctionRef.h"
 #include "mozilla/LinkedList.h"
 #include "nsStyleStruct.h"
 #include "nsCSSPseudoElements.h"
 #include "nsTextNode.h"
+#include <functional>
 
 class nsGenConList;
 class nsIFrame;
@@ -95,6 +97,13 @@
   // Return true if |aNode1| is after |aNode2|.
   static bool NodeAfter(const nsGenConNode* aNode1, const nsGenConNode* aNode2);
 
+  // Find the first element in the list for which the given comparator returns
+  // true. This does a binary search on the list contents.
+  nsGenConNode* BinarySearch(
+      const mozilla::FunctionRef<bool(nsGenConNode*)>& aIsAfter);
+
+  nsGenConNode* GetLast() { return mList.getLast(); }
+
   bool IsFirst(nsGenConNode* aNode) {
     MOZ_ASSERT(aNode, "aNode cannot be nullptr!");
     return aNode == mList.getFirst();